packages/storage/amplify_storage_s3_dart/test/storage_s3_service/task/s3_upload_task_test.dart

```diff
@@ -0,0 +1,902 @@
+
+    group('Uploading S3DataPayload', () {
+      final testDataPayload = S3DataPayload.string('Upload me please!');
+      const testKey = 'object-upload-to';
+
+      test('should invoke S3Client.putObject API with expected parameters',
+          () async {
+        const testUploadDataOptions = S3StorageUploadDataOptions(
+          storageAccessLevel: StorageAccessLevel.private,
+        );
+        final testPutObjectOutput = s3.PutObjectOutput();
+
+        when(
+          () => s3Client.putObject(any()),
+        ).thenAnswer((_) async => testPutObjectOutput);
+
+        final uploadDataTask = S3UploadTask.fromDataPayload(
+          testDataPayload,
+          s3Client: s3Client,
+          prefixResolver: testPrefixResolver,
+          bucket: testBucket,
+          key: testKey,
+          options: testUploadDataOptions,
+          logger: logger,
+          transferDatabase: transferDatabase,
+        );
+
+        unawaited(uploadDataTask.start());
+
+        final result = await uploadDataTask.result;
+
+        expect(result.key, testKey);
+
+        final capturedRequest = verify(
+          () => s3Client.putObject(captureAny<s3.PutObjectRequest>()),
+        ).captured.last;
+
+        expect(capturedRequest is s3.PutObjectRequest, isTrue);
+        final request = capturedRequest as s3.PutObjectRequest;
+        expect(request.bucket, testBucket);
+        expect(
+          request.key,
+          '${await testPrefixResolver.resolvePrefix(
+            storageAccessLevel: testUploadDataOptions.storageAccessLevel,
+          )}$testKey',
+        );
+        expect(request.body, testDataPayload);
+      });
+
+      test(
+          'should invoke S3Client.headObject API with correct parameters when getProperties is set to true in the options',
+          () async {
+        const testUploadDataOptions = S3StorageUploadDataOptions(
+          storageAccessLevel: StorageAccessLevel.private,
+          getProperties: true,
+        );
+        final testPutObjectOutput = s3.PutObjectOutput();
+        final testHeadObjectOutput = s3.HeadObjectOutput();
+
+        when(
+          () => s3Client.putObject(any()),
+        ).thenAnswer((_) async => testPutObjectOutput);
+
+        when(
+          () => s3Client.headObject(any()),
+        ).thenAnswer((_) async => testHeadObjectOutput);
+
+        final uploadDataTask = S3UploadTask.fromDataPayload(
+          testDataPayload,
+          s3Client: s3Client,
+          prefixResolver: testPrefixResolver,
+          bucket: testBucket,
+          key: testKey,
+          options: testUploadDataOptions,
+          logger: logger,
+          transferDatabase: transferDatabase,
+        );
+
+        unawaited(uploadDataTask.start());
+        await uploadDataTask.result;
+
+        final capturedRequest = verify(
+          () => s3Client.headObject(captureAny<s3.HeadObjectRequest>()),
+        ).captured.last;
+
+        expect(capturedRequest is s3.HeadObjectRequest, isTrue);
```

Comment:
non-blocking nit/suggestion: A lot of these requests have repetitive and verbose expects. It might be nice to have some utility/matcher for them but I don't have a more concrete suggestion. Not needed in this PR, just an observation.